### PR TITLE
replace django-tagging by django-tagulous

### LINKED ISSUE #1812: Replace django-tagging with somethings else, i.e. django-tagulous
**Current situation**
Currently DD is using `django-tagging` for tagging products and findings: https://github.com/Fantomas42/django-tagging

I would like to replace this, because:
- it's a performance bottleneck. it's using raw queries underneath, so we can't use `select_related` and `prefetch_related` to have decent performance when rendering lists of products/findings (N+1 queries problem)
- it's old and maybe no longer maintained (latest release Oct '17)
- it's missing features that at some point might be interesting, such as autocomplete.
- tags are not deleted for tests/engagements/products that are deleted, so stuff remains behind in the database 

I'm working on optimizing the queries for the product list page, finding list page, product type list page and active engement page. Most of the queries can be optimized, except the ones related to tags/django-tagging.

**Describe the solution you'd like**
At first I was looking at `django-tagging` but it looks like 'django-tagulous' is the best combination of features and ease of use: https://github.com/radiac/django-tagulous

It's using django models and foreign keys, so everything related to that works as usual, including select_related, prefetch_related and Prefetch.

There is example pseudo code to use django migrations to migrate away from django-tagging.

Some more info: http://radiac.net/projects/django-tagulous/documentation/introduction/#comparison-with-other-tagging-libraries

**Next steps**
Before implementing this I would like some feedback on whether this sounds like a good idea, possible alternatives and of course the chances you would approve to merge this into dev after review/testing.

### LINKED ISSUE #2569: Allow to filter / search by tags
**Is your feature request related to a problem? Please describe**
A clear and concise description of what the problem is.
Ex: I'm always frustrated when I use the view engagements and the issue is I do have tags linked to all engagements.

**Describe the solution you'd like**
As a security analyst, I would like to add available tags as a search option in viewengagements_all. While I realize this is on simple search, we would like to add assigned tags to the view. 

**Describe alternatives you've considered**
A clear and concise description of any alternative solutions or features you have considered.
I have tried to figure this out on my own but I am extremely frustrated. We would rather have this as a display option as opposed to having to export everything to CSV and throw that over to excel. 

**Additional context**
Add any other context, screenshots, sketch, code snippet, etc. about the feature request here.

Add Tags as option like a drop down list. Perhaps there is a development guide somewhere?

image

### LINKED ISSUE #2993: Search product by tags
Hello Team,
We are evaluating DefectDojo in our organization and have found tags to be a very useful feature. We would need to be able to search for a product or an engagement with their tags. It's possible ?

Thanks
Sal

### PR DESCRIPTION
**Pull request by valentijnscholten**
This PR fixes #1812, replacing `django-tagging` by `django-tagulous`. As well as #2993, #2569, #1849.

DONE tasks:
- create unit tests:
- test for finding:
		create with tags
		list and check tags
		get tags
		set tags
		/findings/{id}/tags/
			get, post (post adds to existing)
		/findings/{id}/remove_tags/			
			put, patch (both put and patch remove provided tags from existing, but throw error if tag was not already there)
		tests around spaces and comma's and quotes in tags
- test for import/reimport:
		set tags
- adjust prefetching
- update ui everywhere with new autocomplete field
- update finding bulk edit
- update apply template to finding
- add filtering on tags to UI (fixes #2993)
- update searching by tags (fixes #2569) 
- troubleshoot and report bug to tagulous around fixtures
- rename `Objects` model because it is a reserved keyword/attribute resulting problems with many-to-many relationships
- find out "how the #@$# can we get the width of the tags field correct"
- force lowercase (as was the case with django-tagging)
- update fixtures demodata, testdata, initial
- update serializers for REST API
- add filtering by tag to REST API (#1849)
- set autocomplete / initial for non modelforms ie forms with a tags field
- database migration
- work around `django-tagulous` not supporting serialization to `celery`
- use autocomplete field in filters
- support filtering on multiple tags in APIv2
- make notifications asynchronous again (had to be set to synchronous due to serialization problems)

FIXED some unrelated bugs along the way:
- jira edit engagement not working without JIRA
- add/edit tags for findings wasn't always working
- edit endpoint was not working for invalid data
- bring back in sync the reimport logic in the UI with APIv2

OBSERVATIONS (nothing changed)
- engagement_all page filters on products not on engagements
- bulk edit overwrites existing tags

### CODE REVIEW

File: dojo/templates/dojo/apply_finding_template_form_fields.html [hunk 1]
```diff
@@ -23,7 +23,7 @@
...
+                <select class="form-control" data-field-value="{{ finding_tags|as_json }}" data-template-value="{{ field.value|as_json }}" onchange="(function(t){ var field = document.forms['apply_template']['{{ field.name }}']; var fieldValue = ''; switch(t.value.toLowerCase()){ case 'keep': fieldValue=JSON.parse(t.getAttribute('data-field-value')); break; case 'replace': fieldValue= JSON.parse(t.getAttribute('data-template-value')); break; case 'combine': fieldValue= JSON.parse(t.getAttribute('data-field-value')).concat(JSON.parse(t.getAttribute('data-template-value'))); break;  } $(field).val(fieldValue); })(this)">
```
[the remark this answers is not in the document]
Comment (valentijnscholten) [reply dojo/templates/dojo/apply_finding_template_form_fields.html:32]: pre existing form logic and don't want to refactor too much here.

File: dojo/templates/dojo/apply_finding_template_form_fields.html [hunk 2]
```diff
@@ -23,7 +29,7 @@
...
```
[the remark this answers is not in the document]
Comment (valentijnscholten) [reply dojo/templates/dojo/apply_finding_template_form_fields.html:32]: pre existing form logic and don't want to refactor too much here.

**Review by madchap — APPROVED** [review]
Major work! Applied on my staging env with real data. Migration went smoothly, nothing seems to break.

**Review by mtesauro — APPROVED** [review]
Approve

**Pull request merged**